src/render/optixdenoiser.cpp

```diff
@@ -0,0 +1,391 @@
+#include <mitsuba/render/optixdenoiser.h>
+#include <mitsuba/render/optix_api.h>
+#include <drjit-core/optix.h>
+
+NAMESPACE_BEGIN(mitsuba)
+
+template <typename Float, typename Spectrum>
+static OptixImage2D optixImage2DfromTensor(
+    const typename OptixDenoiser<Float, Spectrum>::TensorXf &tensor,
+    OptixPixelFormat pixel_format) {
+    return { (CUdeviceptr) tensor.data(),
+             (unsigned int) tensor.shape(1),
+             (unsigned int) tensor.shape(0),
+             (unsigned int) (tensor.shape(1) * tensor.shape(2) * sizeof(float)),
+             (unsigned int) (tensor.shape(2) * sizeof(float)),
+             pixel_format };
+}
+
+MI_VARIANT OptixDenoiser<Float, Spectrum>::OptixDenoiser(
+    const ScalarVector2u &input_size, bool albedo, bool normals, bool temporal)
+    : m_input_size(input_size), m_options({ albedo, normals }),
+      m_temporal(temporal) {
+    if constexpr (!dr::is_cuda_v<Float>)
+        Throw("OptixDenoiser is only available in CUDA mode!");
+
+    if (normals && !albedo)
+        Throw("The denoiser cannot use normals to guide its process without "
+              "also providing albedo information!");
+
+    optix_initialize();
+
+    init(input_size, temporal);
+}
+
+MI_VARIANT
+OptixDenoiser<Float, Spectrum>::OptixDenoiser(const OptixDenoiser &other)
+    : m_input_size(other.m_input_size), m_options(other.m_options),
+      m_temporal(other.m_temporal) {
+    init(other.m_input_size, other.m_temporal);
+}
+
+MI_VARIANT
+OptixDenoiser<Float, Spectrum>::OptixDenoiser(OptixDenoiser &&other) noexcept {
```

Comment:
Do we really need to define the move/copy constructor/assignment operator?

Reply:
I doubt anybody is actually going to use them.

However, the implicit default copy operators would be wrong. The other option is to `= delete` them all.

Reply:
Yes, I think it would be enough to use `= delete` here. Anyway, the denoiser inherits from `Object` and will be use with `ref<>` most of the time, in which case we are always moving around a pointer to the denoiser and never really copying/moving its internal data.